external/swloop/swloop.cpp

```diff
@@ -0,0 +1,533 @@
+#include <complex>
+#include "mex.h"
+#include "Eigen/Core"
+#include "Eigen/Cholesky"
+#include "Eigen/Eigenvalues"
+#include <iostream>
+#include <vector>
+#include <thread>
+#include <atomic>
+
+typedef Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> RowMatrixXd;
+typedef std::complex<double> cd_t;
+typedef std::tuple<std::complex<double>, Eigen::VectorXcd> ev_tuple_t;
+
+struct pars {
+    bool hermit;
+    bool formfact;
+    bool incomm;
+    bool helical;
+    bool bq;
+    bool field;
+    double omega_tol;
+    size_t nMagExt;
+    size_t nTwin;
+    size_t nHkl;
+    size_t nBond;
+    size_t nBqBond;
+};
+
+struct swinputs {
+    const double *hklExt;
+    const std::complex<double> *ABCD;
+    const double *idxAll;
+    const double *ham_diag;
+    const double *dR;
+    const double *RR;
+    const double *S0;
+    const std::complex<double> *zed;
+    const double *FF;
+    const double *bqdR;
+    const double *bqABCD;
+    const double *idxBq;
+    const double *bq_ham_diag;
+    std::vector< const double *> ham_MF_v;
+    const int *idx0;
+    const double *n;
+    const double *rotc;
+};
+
+struct swoutputs {
+    std::complex<double> *omega;
+    std::complex<double> *Sab;
+    bool warn_posdef;
+    bool warn_orth;
+};
+
+template <typename T>
+using MatrixXT = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>;
+
+template <typename T>
+MatrixXT<T> accumarray(Eigen::MatrixXd ind,
+                       Eigen::Array<T, Eigen::Dynamic, 1> data,
+                       size_t sz1, size_t sz2=0) {
+    // Note this version of accumarray is explicitly 2D
+    if (sz2 == 0)
+        sz2 = sz1;
+    MatrixXT<T> res = MatrixXT<T>::Zero(sz1, sz2);
+    for (int ii=0; ii<ind.rows(); ii++) {
+        // Indices are direct from Matlab so use base-1 and are also doubles not ints
+        res(static_cast<size_t>(ind(ii, 0)) - 1, static_cast<size_t>(ind(ii, 1)) - 1) += data(ii);
+    }
+    return res;
+}
+
+std::atomic_int err_flag(0);
+std::vector<std::string> errmsgs = {
+    std::string("all good"),
+    std::string("swloop:notposdef: Very baaaad!"),
```

Comment:
Duplicate message here? If they refer to different matrices being not positive definite could we specify which?

Reply:
It's to duplicate the behaviour in [spinwave.m](https://github.com/SpinW/spinw/blob/692683efa71076405654fa5f6b63585f12a0afe6/swfiles/%40spinw/spinwave.m#L822) where it throws the non-positive definite error at two different points with different error messages. It's probably best to clean up both the `.m` and `mex` codes with respects to this...